apiserver/facades/agent/uniter/uniter.go

```diff
@@ -403,11 +466,56 @@ func (u *UniterAPI) AllMachinePorts(args params.Entities) (params.MachinePortsRe
 		return params.MachinePortsResults{}, err
 	}
 	for i, entity := range args.Entities {
-		result.Results[i] = u.getOneMachinePorts(canAccess, entity.Tag)
+		machPortRanges, err := u.getOneMachineOpenedPortRanges(canAccess, entity.Tag)
+		if err != nil {
+			result.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+
+		// This method is used by older clients that always open ports
+		// to all subnets; to emulate this behavior, we get the unique
+		// set of ports for each unit and return them back as a sorted
+		// list.
+		var resultPorts []params.MachinePortRange
+		for unitName, unitPortRanges := range machPortRanges.ByUnit() {
+			unitTag := names.NewUnitTag(unitName).String()
+			for _, pr := range unitPortRanges.UniquePortRanges() {
+				resultPorts = append(resultPorts, params.MachinePortRange{
+					UnitTag:   unitTag,
+					PortRange: params.FromNetworkPortRange(pr),
+				})
+			}
+		}
+
+		// Sort result by port range to ensure stable order for returned ranges.
+		sort.Slice(resultPorts, func(i, j int) bool {
+			return resultPorts[i].PortRange.NetworkPortRange().LessThan(
+				resultPorts[j].PortRange.NetworkPortRange(),
+			)
+		})
+
+		result.Results[i] = params.MachinePortsResult{
+			Ports: resultPorts,
+		}
 	}
 	return result, nil
 }
 
+func (u *UniterAPI) getOneMachineOpenedPortRanges(canAccess common.AuthFunc, machineTag string) (state.MachinePortRanges, error) {
+	tag, err := names.ParseMachineTag(machineTag)
+	if err != nil {
+		return nil, apiservererrors.ErrPerm
+	}
+	if !canAccess(tag) {
+		return nil, apiservererrors.ErrPerm
```

Comment:
It'd be helpful annotate these ErrPerm, so we know which ErrPerm it is when tracking it down.

Reply:
Just following the pattern seen in every other method here :D 

However, well behaving clients (i.e. the uniter as I don't think pylibjuju should be messing with these APIs) should never hit an ErrPerm. I could error.Trace them although I am not entirely sure whether the stack gets propagated when you convert into a ServerError.